Engine/Filesystem/Path.cpp

```diff
@@ -0,0 +1,292 @@
+	is_directory = (PHYSFS_FileType::PHYSFS_FILETYPE_DIRECTORY == path_info.filetype);
+	CalculateFile();
+}
+
+void Path::CalculateFile()
+{
+	if (is_directory)
+	{
+		file = nullptr;
+	}
+	else
+	{
+		file = new File(this);
+	}
+}
+
+void Path::CalculateChildren()
+{
+	std::vector<Path*> path_children;
+	GetAllFilesInPath(path_children);
+
+	for (auto & path_child : path_children)
+	{
+		path_child->parent = this;
+		this->children.push_back(path_child);
+		if (path_child->IsDirectory())
+		{
+			++sub_folders;
+			total_sub_files_number += path_child->total_sub_files_number;
+		}
+		else
+		{
+			++total_sub_files_number;
+		}
+	}
+}
+
+uint32_t Path::GetModificationTimestamp() const
+{
+	PHYSFS_Stat path_info;
+	if (PHYSFS_stat(file_path.c_str(), &path_info) == 0)
+	{
+		APP_LOG_ERROR("Error getting %s path info: %s", file_path.c_str(), PHYSFS_getLastError());
+	}
+
+	return path_info.modtime;
+}
+
+void Path::GetAllFilesInPath(std::vector<Path*>& path_children, bool directories_only)
+{
+	char **files_array = PHYSFS_enumerateFiles(file_path.c_str());
+	if (*files_array == NULL)
+	{
+		APP_LOG_INFO("Error reading directory %s: %s", file_path.c_str(), PHYSFS_getLastError());
+		return;
+	}
+
+	char **filename;
+	for (filename = files_array; *filename != NULL; filename++)
+	{
+		Path* child_path = new Path(file_path, *filename);
+		if ((directories_only && child_path->is_directory) || !directories_only)
+		{
+			path_children.push_back(child_path);
+		}
+	}
+	PHYSFS_freeList(files_array);
+}
+
+bool Path::IsDirectory() const
+{
+	return is_directory;
+}
+
+bool Path::IsMeta() const
+{
+	return !is_directory && file->GetFileType() == FileType::META;
+}
+
+bool Path::IsImportable() const
+{
+	FileType file_type = file->GetFileType();
+	return
+		file_type == FileType::ANIMATION
+		|| file_type == FileType::MATERIAL
+		|| file_type == FileType::MESH
+		|| file_type == FileType::MODEL
+		|| file_type == FileType::PREFAB
+		|| file_type == FileType::SKELETON
+		|| file_type == FileType::SKYBOX
+		|| file_type == FileType::STATE_MACHINE
+		|| file_type == FileType::TEXTURE;
+}
+
+std::string Path::GetFullPath() const
+{
+	return file_path;
+}
+
+std::string Path::GetFilename() const
+{
+	return file_path.substr(file_path.find_last_of('/') + 1, -1);
+}
+
+std::string Path::GetFilenameWithoutExtension() const
+{
+	std::string file_name = GetFilename();
+	return file_name.substr(0, file_name.find_last_of("."));
+}
+
+std::string Path::GetExtension() const
+{
+	std::size_t found = file_path.find_last_of(".");
+	if (found == std::string::npos || found == 0) {
+		return "";
+	}
+	std::string file_extension = file_path.substr(found + 1, file_path.length());
+
+	return file_extension;
+}
+
+std::string Path::GetExtension(const std::string& path)
+{
+	std::size_t found = path.find_last_of(".");
+	if (found == std::string::npos || found == 0) {
+		return "";
+	}
+	std::string file_extension = path.substr(found + 1, path.length());
+
+	return file_extension;
+}
```

Comment:
Maybe this function can be const?

Reply:
Yeah, probably! Good catch

Reply:
In fact no, because it's a static function, and static functions cannot be const